### PR TITLE
lib: intro the matrix package

### PR DESCRIPTION
**Pull request by xymus**
This PR intro `Matrix` to represent a matrix of floats. The main module offers only general services and the `projection` module adds services useful for 3D manipulations.

The package is oriented towards matrices of floats and was designed alongside Gamnit. As such, is a bit specialized and may not be appropriate for all usage of matrices (such as integer matrices). So if you prefer, I can move these back to Gamnit. But since this module does not have any dependencies and `Matrix` can be useful for a wide array of context, I chose to leave it in lib.

### CODE REVIEW

File: lib/matrix/matrix.nit [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+
+	# Number of columns
+	var width: Int
+
+	# Number of rows
+	var height: Int
+
+	# Items of this matrix, rows by rows
+	private var items: Array[Float] is lazy	do
+		return [for i in (width*height).times do 0.0]
```
Comment (privat) [inline lib/matrix/matrix.nit]:
~~~
return new Array[Float].filled_with(width*height, 0.0)
return [0.0] * (width*height)
~~~

File: lib/matrix/matrix.nit [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	end
+
+	# Create a matrix from nested sequences
+	#
+	# Require: `not items.is_empty`
```
Comment (privat) [inline lib/matrix/matrix.nit]: For robustness, shoud not an empty items return an empty matrix?

File: lib/matrix/matrix.nit [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	# Require: all rows are of the same length
+	#
+	# ~~~
+	# var matrix = new Matrix.from([[1.0, 2.0],
+	#                               [3.0, 4.0]])
+	# assert matrix.to_s == """
+	# 1.0 2.0
+	# 3.0 4.0"""
+	# ~~~
+	init from(items: SequenceRead[SequenceRead[Float]])
```
Comment (privat) [inline lib/matrix/matrix.nit]: I propose to add a reciprocal operation that produces a `Array[Array[Float]]`

File: lib/matrix/matrix.nit [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+	# ~~~
+	# var i = new Matrix.identity(3)
+	# assert i.to_s == """
+	# 1.0 0.0 0.0
+	# 0.0 1.0 0.0
+	# 0.0 0.0 1.0"""
+	# ~~~
+	new identity(size: Int)
+	do
+		var matrix = new Matrix(size, size)
```
Comment (privat) [inline lib/matrix/matrix.nit]: assert size >= 0

File: lib/matrix/projection.nit [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+
+		return new Matrix.frustum(-frustum_width, frustum_width,
+		                          -frustum_height, frustum_height,
+		                          near, far)
+	end
+
+	# Create a frustum transformation matrix
+	#
+	# `left, right, bottom, top, near, far` defines the world clip planes.
+	new frustum(left, right, bottom, top, near, far: Float)
```
Comment (privat) [inline lib/matrix/projection.nit]: wat? Maybe add a link to wikipedia :)

**Comment by privat** [comment]
+1
You can also add a `transpose` method that is quite standard, if you want

File: lib/matrix/matrix.nit (hunk 3 above)
Comment (xymus) [inline lib/matrix/matrix.nit]: Added as `to_a`.

File: lib/matrix/projection.nit (hunk 5 above)
Comment (xymus) [inline lib/matrix/projection.nit]: It's the matrix used to create the perspective effect in 3D games.

I prefer not linking to wikipedia in my doc, I thrust that the users know how to use Google!

**Comment by xymus** [comment]
Rerolled with small fixes and to add `to_a` and `transposed`.

**Pull request merged**